Sources/MongoSwift/BSON/Document+Collection.swift

```diff
@@ -5,43 +5,46 @@ import Foundation
 /// ordering to the key-value pairs in the document.
 extension Document: Collection {
     /// Returns the start index of the Document.
-    public var startIndex: Int {
+    public var startIndex: Index {
         return 0
     }
 
     /// Returns the end index of the Document.
-    public var endIndex: Int {
+    public var endIndex: Index {
         return self.count
     }
 
-    private func failIndexCheck(_ i: Int) {
+    private func failIndexCheck(_ i: Index) {
         let invalidIndexMsg = "Index \(i) is invalid"
         guard !self.isEmpty && self.startIndex ... self.endIndex - 1 ~= i else {
             fatalError(invalidIndexMsg)
         }
     }
 
     /// Returns the index after the given index for this Document.
-    public func index(after i: Int) -> Int {
+    public func index(after i: Index) -> Index {
         // Index must be a valid one, meaning it must exist somewhere in self.keys.
         failIndexCheck(i)
         return i + 1
     }
 
     /// Allows access to a `KeyValuePair` from the `Document`, given the position of the desired `KeyValuePair` held
     /// within. This method does not guarantee constant-time (O(1)) access.
-    public subscript(position: Int) -> KeyValuePair {
+    public subscript(position: Index) -> KeyValuePair {
         // TODO: This method _should_ guarantee constant-time O(1) access, and it is possible to make it do so. This
         // criticism also applies to key-based subscripting via `String`.
         // See SWIFT-250.
         failIndexCheck(position)
+
+        let subsequence = DocumentIterator.subsequence(of: self, startIndex: position, endIndex: position + 1)
```

Comment:
this might be a little clearer if we introduced an `init(forDocument: Document, advancedToIndex: Index)` to `DocumentIterator`. It would also mirror how we get an individual element via a string key.

Reply:
How would that be used here?

Reply:
we'd be able to do:
```swift
let iter = DocumentIterator(forDocument: self, advancedToIndex: position)
return (iter.currentKey, iter.currentValue)
```
As a side note, `currentKeyValuePair` may be a nice computed property on `DocumentIterator` so we could do it in one line:
```swift
return DocumentIterator(forDocument: self, advancedToIndex: position).currentKeyValuePair
```

Reply:
I like the first one! I'll add it now.

Reply:
I ended up just creating a `DocumentIterator` and advancing it it (this is what the convenience init would have done). Using the custom initializer here actually made it difficult to reason about all the potential error cases (should we want to test for all of them):
  - `failIndexCheck` is happening in the subscript, but it _should_ happen in the initializer
  - should the initializer catch an error on `iter.advance()`?
  - it's not even clear to me how one guards a failable initializer, this code does not work:
```
internal convenience init?(forDocument doc: Document, advancedToIndex pos: Document.Index) {
    guard let me = self.init(forDocument: doc) else {
        return nil
    }
    ...
}
```

We only use this in one place, so I chose instead to localize the logic at the call site and not overeagerly DRY the code.